### PR TITLE
improve CPU performance for log_softmax when dim != -1 on both float32 and bfloat16

### PR DESCRIPTION
**Pull request by mingfeima**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64726
* #64479
* #64478

### CODE REVIEW

**Comment by mingfeima** [comment]
`log_softmax` is the major bottleneck in `nn.CrossEntropyLoss`. The original kernel has fast path for case when `dim = -1` but will go reference path when `dim != -1`.

This patch optimized log_softmax when dim != -1.

The algorithm of log softmax will read the input 3 times: a) compute max; b) compute sum(x - max); c) compute x - max - sum
So this implementation did blocking on inner_size so as to increase cache hit rate. Each block takes 128KB which is L2 cache fit for both old arch (Haswell, L2 256KB) and more modern arch (Skylake, Cascade Lake and later: L2 1MB).

On AVX512 CPU:
* Single socket run receives 5.4x (fp32) and 20.0x (bf16) perf boost.
* Single core sun receives 7.7x (fp32) and 25.0x (bf16) perf boost.

Benchmark shows the optimization is equally effective on both avx2 as avx512 CPU.

#### avx512 CPU single socket
```bash
### before
##  LogSoftmax dim = 1 ; size:  torch.Size([32, 81, 8732]) ; dtype:  torch.float32 ; time = 18.058 ms
##  LogSoftmax dim = 1 ; size:  torch.Size([32, 81, 8732]) ; dtype:  torch.bfloat16 ; time = 47.020 ms

### after
##  LogSoftmax dim = 1 ; size:  torch.Size([32, 81, 8732]) ; dtype:  torch.float32 ; time = 3.297 ms
##  LogSoftmax dim = 1 ; size:  torch.Size([32, 81, 8732]) ; dtype:  torch.bfloat16 ; time = 2.342 ms
```
#### avx512 CPU single core
```bash
### before
##  LogSoftmax dim = 1 ; size:  torch.Size([32, 81, 8732]) ; dtype:  torch.float32 ; time = 302.236 ms
##  LogSoftmax dim = 1 ; size:  torch.Size([32, 81, 8732]) ; dtype:  torch.bfloat16 ; time = 777.798 ms

### after
##  LogSoftmax dim = 1 ; size:  torch.Size([32, 81, 8732]) ; dtype:  torch.float32 ; time = 39.355 ms
##  LogSoftmax dim = 1 ; size:  torch.Size([32, 81, 8732]) ; dtype:  torch.bfloat16 ; time = 31.144 ms
```
#### avx2 CPU single socket
```bash
### before
##  LogSoftmax dim = 1 ; size:  torch.Size([32, 81, 8732]) ; dtype:  torch.float32 ; time = 66.368 ms
##  LogSoftmax dim = 1 ; size:  torch.Size([32, 81, 8732]) ; dtype:  torch.bfloat16 ; time = 94.870 ms

### after
##  LogSoftmax dim = 1 ; size:  torch.Size([32, 81, 8732]) ; dtype:  torch.float32 ; time = 7.894 ms
##  LogSoftmax dim = 1 ; size:  torch.Size([32, 81, 8732]) ; dtype:  torch.bfloat16 ; time = 6.726 ms
```

#### avx2 CPU single core
```bash
### before
##  LogSoftmax dim = 1 ; size:  torch.Size([32, 81, 8732]) ; dtype:  torch.float32 ; time = 590.100 ms
##  LogSoftmax dim = 1 ; size:  torch.Size([32, 81, 8732]) ; dtype:  torch.bfloat16 ; time = 1045.081 ms

### after
##  LogSoftmax dim = 1 ; size:  torch.Size([32, 81, 8732]) ; dtype:  torch.float32 ; time = 64.109 ms
##  LogSoftmax dim = 1 ; size:  torch.Size([32, 81, 8732]) ; dtype:  torch.bfloat16 ; time = 50.494 ms
```

**Review by frank-wei — APPROVED** [review]
Passed tests on D33862416
